pkg/controller/expcertificates/issuing/issuing_controller.go

```diff
@@ -0,0 +1,396 @@
+
+	reqs, nextRevision, err := findNextRevisionCertificateRequestsForCertificate(log, crt, c.certificateRequestLister)
+	if err != nil {
+		return err
+	}
+
+	if len(reqs) != 1 {
+		// If none exist do nothing. If multiple exist, then leave to
+		// requestmanager controller to clean up.
+		return nil
+	}
+
+	req := reqs[0]
+
+	reqReason := apiutil.CertificateRequestReadyReason(req)
+	switch reqReason {
+
+	// If the certificate request has failed, set the last failure time to now,
+	// and set the Issuing status condition to False with reason.
+	case cmapi.CertificateRequestReasonFailed:
+		nowTime := metav1.NewTime(c.clock.Now())
+		crt.Status.LastFailureTime = &nowTime
+
+		var reason, message string
+
+		reason = "FailedToIssue"
+		message = fmt.Sprintf("The certificate request has failed to issue and will be retired: %s",
+			apiutil.CertificateRequestFailedMessage(req))
+
+		crt = crt.DeepCopy()
+		apiutil.SetCertificateCondition(crt, cmapi.CertificateConditionIssuing, cmmeta.ConditionFalse, reason, message)
+
+		if err := c.updateCertificateStatus(ctx, crt); err != nil {
+			return err
+		}
+
+		c.recorder.Event(crt, corev1.EventTypeWarning, "Issuing", message)
+
+		return nil
+
+		// If the CertificateRequest is valid, verify its status and update accordingly.
+	case cmapi.CertificateRequestReasonIssued:
+
+		// Verify the public key of the certificate request matches the public key of the stored status.nextPrivateKeySecretName
+		publicKeyMatches, err := c.certificateRequestPublicKeyMatchesSecret(ctx, req, namespace, nextPrivateKeySecretName)
+		if err != nil {
+			return err
+		}
+
+		// If public key does not match, do nothing (requestmanager will handle this).
+		if !publicKeyMatches {
+			return nil
+		}
+
+		// Verify the CSR options match what is requested in certificate.spec.
+		violations, err := certificates.RequestMatchesSpec(req, crt.Spec)
+		if err != nil {
+			return err
+		}
+
+		// If there are violations in the spec, then the requestmanager will handle this.
+		if len(violations) > 0 {
+			return nil
+		}
+
+		//Encode and issue the key-pair (store it in the Secret)
+		if err := c.issueCertificateAndKeyToSecret(ctx, namespace, crt, req); err != nil {
+			return err
+		}
+
+		//Set status.revision to revision of the CertificateRequest
+		crt.Status.Revision = &nextRevision
+
+		crt = crt.DeepCopy()
+
+		// Remove Issuing status condition
+		apiutil.RemoveCertificateCondition(crt, cmapi.CertificateConditionIssuing)
+
+		//Clear status.lastFailureTime (if set)
+		crt.Status.LastFailureTime = nil
+
+		if err := c.updateCertificateStatus(ctx, crt); err != nil {
+			return err
+		}
+
+		message := "The certificate has been successfully issued"
+		c.recorder.Event(crt, corev1.EventTypeNormal, "Issuing", message)
+
+		return nil
+
+	// CertificateRequest is not in a final state so do nothing.
+	default:
+		return nil
+	}
+}
+
+func (c *controller) updateCertificateStatus(ctx context.Context, crt *cmapi.Certificate) error {
```

Comment:
Can we remove this function altogether, set a 10s deadline at the *start* of the sync loop, and then just pass `ctx` directly to `UpdateStatus` at the places we currently call this function? It'll ultimately be a reduction in lines of code and also brings the _actual_ function being called directly to where it is used, which helps make things easier to follow :)